### ISSUE TITLE
ptr_arg false positive when Cow's lifetime appears in return type

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust-clippy#9218]
### Summary

Clippy's warn-by-default `ptr_arg` lint is displeased about function arguments with the type `&Cow<T>`. That's fine, but `&Cow<'lifetime, T>` should be exempt from this lint when the Cow's lifetime is not elided.

I hit this false positive in serde_yaml code that deals with deserializing borrowed &amp;str. Since YAML deserializers can be constructed from either a `&'de str` as input (`from_str`) or `std::io::Read` as input (`from_reader`), and only the former supports deserializing output values that borrow from the input data, some stuff is represented using a `Cow` internally which is `Borrowed` if deserializing from str and `Owned` if deserializing from reader.

### Lint Name

ptr_arg

### Reproducer

```rust
use std::borrow::Cow;

fn repro<'a>(input: &Cow<'a, str>) -> (usize, Option<&'a str>) {
    let wow = input.len() / 2;
    let huh = match input {
        Cow::Borrowed(borrowed) => Some(&borrowed[wow..]),
        Cow::Owned(_) => None,
    };
    (wow, huh)
}

fn main() {
    let input = Cow::Borrowed("asdfjkl");
    assert_eq!(repro(&input), (3, Some("fjkl")));

    let input = Cow::Owned("asdfjkl".to_owned());
    assert_eq!(repro(&input), (3, None));
}
```

```console
$ cargo clippy
warning: using a reference to `Cow` is not recommended
 --> src/main.rs:3:21
  |
3 | fn repro<'a>(input: &Cow<'a, str>) -> (usize, Option<&'a str>) {
  |                     ^^^^^^^^^^^^^ help: change this to: `&str`
  |
  = note: `#[warn(clippy::ptr_arg)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_arg

warning: `testing` (bin "testing") generated 1 warning
```

### Version

```text
rustc 1.64.0-nightly (9a7b7d5e5 2022-07-19)
binary: rustc
commit-hash: 9a7b7d5e50ab0b59c6d349bbf005680a7c880e98
commit-date: 2022-07-19
host: x86_64-unknown-linux-gnu
release: 1.64.0-nightly
LLVM version: 14.0.6
```

**Comment by Diggsey** [comment]
I also ran into this same issue 👍
